### PR TITLE
ADCM-2981 Operation and Login objects for Audit

### PR DESCRIPTION
**Pull request by Sealwing**

### CODE REVIEW

File: src/adcm_client/util/search.py [hunk 1]
```diff
@@ -89,14 +89,20 @@ def filter_data(x):
...
+        result = [k in new_data and v == new_data[k] for k, v in new_attrs.items()]
         if result:
             return all(result)
+        # Why?
+        # `result` will be empty if nothing matched `new_attrs`,
+        # but in case of filters that don't have corresponding fields in object's body
+        # we've got them filtered out! (e.g. `operation_date` for audit log)
+        # Maybe we want to pass checks if `new_attrs` is empty?
+        # Or there's motivation for not doing that?
         return False
```
Comment (Sealwing) [inline src/adcm_client/util/search.py:106]: @a-alferov Can you, please, check out this section and especially the comment?

File: src/adcm_client/audit.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    IDNAME = 'id'
+    PATH = ['audit', 'login']
+    FILTERS = ['login_result', 'login_date', 'username']
+
+    id: int = None
+    user_id: int = None
+    login_result: LoginResult = None
+    login_time: datetime.datetime = None
+
```
Comment (Sealwing) [inline src/adcm_client/audit.py:133]: `details` missing

**Comment by Sealwing** [comment]
bender build

File: src/adcm_client/base.py [hunk 3]
```diff
@@ -443,3 +458,103 @@ def __init__(self, api: ADCMApiWrapper, path=None, path_args=None, paging=None,
...
+        # but since it's an int, there shouldn't be a problem anyway
+        kwargs.update({k: _simplify_filter(v) for k, v in kwargs.items() if k in self.FILTERS})
+        super().__init__(*args, **kwargs)
+
+    def _register_attrs(self):
+        super()._register_attrs()
+        try:
+            self._convert()
+        except Exception as e:  # pylint: disable=broad-except
+            warnings.warn(f'Some of the fields left was not converted due to error: {e}\n')
```
Comment (Sealwing) [inline src/adcm_client/base.py:502]: moved conversion to here from __init__, because of `reread`

File: src/adcm_client/base.py [hunk 4]
```diff
@@ -443,3 +458,103 @@ def __init__(self, api: ADCMApiWrapper, path=None, path_args=None, paging=None,
...
+        """Convert string to datetime in iso format"""
+        raw_value = getattr(self, field)
+        # most likely DRF format
+        try:
+            setattr(
+                self,
+                field,
+                datetime.datetime.strptime(raw_value, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
+                    tzinfo=datetime.timezone.utc
+                ),
```
Comment (Sealwing) [inline src/adcm_client/base.py:529]: In my understanding, the Z on the end means this date is UTC, so we can just set "correct" timezone

File: src/adcm_client/base.py [hunk 5]
```diff
@@ -443,3 +463,101 @@ def __init__(self, api: ADCMApiWrapper, path=None, path_args=None, paging=None,
...
+    def _convert(self):
+        """Explicitly convert each field that has complex type after the data was fetched"""
+        raise NotImplementedError
+
+    def _convert_enum(self, field: str, enum_cls: Type[Enum]):
+        raw_value = getattr(self, field)
+        try:
+            setattr(self, field, enum_cls(raw_value))
+        except ValueError:
+            warnings.warn(
```
Comment (Sealwing) [inline src/adcm_client/base.py:519]: TODO: check if value is None first

File: src/adcm_client/base.py (hunk 3 above)
Comment (a-alferov) [reply src/adcm_client/base.py:507]: Ок

**Review by a-alferov — APPROVED** [review]

**Comment by a-alferov** [comment]
bender build

**Pull request merged**